### ISSUE TITLE
globalkeys daemon sometimes freezes

### ISSUE THREAD

**Issue by RalfJung** [lxde/lxqt#992]
In a couple dozen logins, it now happened twice for me that the `globalkeys` daemon froze. This was noticeable in the following ways:
- Pressing a shortcut did not do anything, until roughly a minute later.
- The panel would also freeze. Attaching gdb to the panel revealed the following backtrace

```
#0  0x00007f0839d1be4d in poll () at ../sysdeps/unix/syscall-template.S:84
#1  0x00007f0838775dff in ?? () from /lib/x86_64-linux-gnu/libdbus-1.so.3
#2  0x00007f0838774a6e in ?? () from /lib/x86_64-linux-gnu/libdbus-1.so.3
#3  0x00007f083875d6c4 in ?? () from /lib/x86_64-linux-gnu/libdbus-1.so.3
#4  0x00007f083875e09d in ?? () from /lib/x86_64-linux-gnu/libdbus-1.so.3
#5  0x00007f083d416815 in q_dbus_pending_call_block (pending=<optimized out>) at qdbus_symbols_p.h:385
#6  QDBusConnectionPrivate::waitForFinished (this=0x19020e0, pcall=pcall@entry=0x1c57ab0) at qdbusintegrator.cpp:1817
#7  0x00007f083d45ace4 in QDBusPendingCallPrivate::waitForFinished (this=0x1c57ab0) at qdbuspendingcall.cpp:234
#8  0x00007f083c967e64 in ?? () from /usr/lib/x86_64-linux-gnu/liblxqt-globalkeys.so.0
#9  0x00007f083c96be7e in ?? () from /usr/lib/x86_64-linux-gnu/liblxqt-globalkeys.so.0
#10 0x00007f083c96c0b1 in GlobalKeyShortcut::Action::changeShortcut(QString const&) () from /usr/lib/x86_64-linux-gnu/liblxqt-globalkeys.so.0
#11 0x000000000045cd79 in DesktopSwitch::shortcutRegistered (this=<optimized out>) at /build/lxqt-panel-Nn8njZ/lxqt-panel-0.10.0/plugin-desktopswitch/desktopswitch.cpp:109
#12 0x00007f083a90de0f in QtPrivate::QSlotObjectBase::call (a=0x7ffe6c033250, r=0x1921dc0, this=<optimized out>) at ../../include/QtCore/../../src/corelib/kernel/qobject_impl.h:124
#13 QMetaObject::activate (sender=0x1b4bee0, signalOffset=<optimized out>, local_signal_index=<optimized out>, argv=<optimized out>) at kernel/qobject.cpp:3698
#14 0x00007f083a90df8a in QMetaObject::activate (sender=0x1b4a9a0, signalOffset=<optimized out>, local_signal_index=<optimized out>, argv=<optimized out>) at kernel/qobject.cpp:3713
#15 0x00007f083c969e67 in ?? () from /usr/lib/x86_64-linux-gnu/liblxqt-globalkeys.so.0
#16 0x00007f083c96ec5c in ?? () from /usr/lib/x86_64-linux-gnu/liblxqt-globalkeys.so.0
#17 0x00007f083a90df8a in QMetaObject::activate (sender=0x1b4c5e0, signalOffset=<optimized out>, local_signal_index=local_signal_index@entry=0, argv=argv@entry=0x7ffe6c0335d0) at kernel/qobject.cpp:3713
#18 0x00007f083a90e797 in QMetaObject::activate (sender=<optimized out>, m=m@entry=0x7f083d4746a0 <QDBusPendingCallWatcher::staticMetaObject>, local_signal_index=local_signal_index@entry=0, argv=argv@entry=0x7ffe6c0335d0)
    at kernel/qobject.cpp:3578
#19 0x00007f083d459f6f in QDBusPendingCallWatcher::finished (this=<optimized out>, _t1=0x1b4c5e0) at .moc/moc_qdbuspendingcall.cpp:148
#20 0x00007f083d45b645 in QDBusPendingCallWatcherPrivate::_q_finished (this=<optimized out>) at qdbuspendingcall.cpp:488
#21 QDBusPendingCallWatcher::qt_static_metacall (_o=<optimized out>, _c=<optimized out>, _id=<optimized out>, _a=<optimized out>) at .moc/moc_qdbuspendingcall.cpp:80
#22 0x00007f083a90ee61 in QObject::event (this=0x1b4c5e0, e=<optimized out>) at kernel/qobject.cpp:1239
#23 0x00007f083b1d205c in QApplicationPrivate::notify_helper (this=this@entry=0x1837af0, receiver=receiver@entry=0x1b4c5e0, e=e@entry=0x1cf38e0) at kernel/qapplication.cpp:3716
#24 0x00007f083b1d7516 in QApplication::notify (this=0x7ffe6c033bd0, receiver=0x1b4c5e0, e=0x1cf38e0) at kernel/qapplication.cpp:3499
#25 0x00007f083a8df5ab in QCoreApplication::notifyInternal (this=0x7ffe6c033bd0, receiver=0x1b4c5e0, event=event@entry=0x1cf38e0) at kernel/qcoreapplication.cpp:965
#26 0x00007f083a8e19a6 in QCoreApplication::sendEvent (event=0x1cf38e0, receiver=<optimized out>) at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:224
#27 QCoreApplicationPrivate::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0, data=0x1836900) at kernel/qcoreapplication.cpp:1593
#28 0x00007f083a8e1e88 in QCoreApplication::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0) at kernel/qcoreapplication.cpp:1451
#29 0x00007f083a935633 in postEventSourceDispatch (s=0x1866980) at kernel/qeventdispatcher_glib.cpp:271
#30 0x00007f083cbbdfd7 in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#31 0x00007f083cbbe230 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#32 0x00007f083cbbe2dc in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#33 0x00007f083a935a3f in QEventDispatcherGlib::processEvents (this=0x1868fd0, flags=...) at kernel/qeventdispatcher_glib.cpp:418
#34 0x00007f083a8dcd6a in QEventLoop::exec (this=this@entry=0x7ffe6c033b60, flags=..., flags@entry=...) at kernel/qeventloop.cpp:204
#35 0x00007f083a8e4e0c in QCoreApplication::exec () at kernel/qcoreapplication.cpp:1229
#36 0x00007f083ac18c1c in QGuiApplication::exec () at kernel/qguiapplication.cpp:1539
#37 0x00007f083b1ce495 in QApplication::exec () at kernel/qapplication.cpp:2976
#38 0x000000000042fa74 in main (argc=1, argv=<optimized out>) at /build/lxqt-panel-Nn8njZ/lxqt-panel-0.10.0/panel/main.cpp:41
```

So it is blocking somewhere in `GlobalKeyShortcut::Action::changeShortcut`. (Doing blocking calls in the event loop thread is asking for trouble, but I guess that's a different bug.)

I tried figuring out where it froze, but could only get the following rather useless backtrace:

```
#0  0x00007f50db7d4e2d in read () at ../sysdeps/unix/syscall-template.S:84
#1  0x0000000000420e20 in ?? ()
#2  0x000000000040815c in ?? ()
#3  0x0000000000408a83 in ?? ()
#4  0x000000000040c51b in ?? ()
#5  0x000000000040e9cd in ?? ()
#6  0x00000000004126ef in ?? ()
#7  0x00000000004277ec in ?? ()
#8  0x00007f50dc2e2f8a in QMetaObject::activate (sender=0x16b6a30, signalOffset=<optimized out>, local_signal_index=<optimized out>, argv=<optimized out>) at kernel/qobject.cpp:3713
#9  0x0000000000426a52 in ?? ()
#10 0x0000000000420405 in ?? ()
#11 0x0000000000425a2b in ?? ()
#12 0x0000000000425ca3 in ?? ()
#13 0x00007f50dd2a4913 in QDBusConnectionPrivate::deliverCall (this=this@entry=0x16a3570, object=object@entry=0x16b0f40, msg=..., metaTypes=..., slotIdx=5) at qdbusintegrator.cpp:978
#14 0x00007f50dd2a87c0 in QDBusConnectionPrivate::activateCall (this=this@entry=0x16a3570, object=0x16b0f40, flags=flags@entry=273, msg=...) at qdbusintegrator.cpp:892
#15 0x00007f50dd2a9073 in QDBusConnectionPrivate::activateObject (this=0x16a3570, node=..., msg=..., pathStartPos=<optimized out>) at qdbusintegrator.cpp:1462
#16 0x00007f50dd2aa26e in QDBusActivateObjectEvent::placeMetaCall (this=0x16c3060) at qdbusintegrator.cpp:1580
#17 0x00007f50dc2e3e61 in QObject::event (this=0x16b6a30, e=<optimized out>) at kernel/qobject.cpp:1239
#18 0x00007f50dc65f05c in QApplicationPrivate::notify_helper (this=this@entry=0x163f160, receiver=receiver@entry=0x16b6a30, e=e@entry=0x16c3060) at kernel/qapplication.cpp:3716
#19 0x00007f50dc664516 in QApplication::notify (this=0x7ffdce99e3d0, receiver=0x16b6a30, e=0x16c3060) at kernel/qapplication.cpp:3499
#20 0x00007f50dc2b45ab in QCoreApplication::notifyInternal (this=0x7ffdce99e3d0, receiver=0x16b6a30, event=event@entry=0x16c3060) at kernel/qcoreapplication.cpp:965
#21 0x00007f50dc2b69a6 in QCoreApplication::sendEvent (event=0x16c3060, receiver=<optimized out>) at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:224
#22 QCoreApplicationPrivate::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0, data=0x163f2d0) at kernel/qcoreapplication.cpp:1593
#23 0x00007f50dc2b6e88 in QCoreApplication::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0) at kernel/qcoreapplication.cpp:1451
#24 0x00007f50dc30a633 in postEventSourceDispatch (s=0x166e4e0) at kernel/qeventdispatcher_glib.cpp:271
#25 0x00007f50d9ec4fd7 in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#26 0x00007f50d9ec5230 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#27 0x00007f50d9ec52dc in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#28 0x00007f50dc30aa3f in QEventDispatcherGlib::processEvents (this=0x16a0df0, flags=...) at kernel/qeventdispatcher_glib.cpp:418
#29 0x00007f50dc2b1d6a in QEventLoop::exec (this=this@entry=0x7ffdce99e2e0, flags=..., flags@entry=...) at kernel/qeventloop.cpp:204
#30 0x00007f50dc2b9e0c in QCoreApplication::exec () at kernel/qcoreapplication.cpp:1229
#31 0x0000000000407139 in ?? ()
#32 0x00007f50db719610 in __libc_start_main (main=0x406b40, argc=1, argv=0x7ffdce99e618, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffdce99e608) at libc-start.c:291
#33 0x00000000004072e9 in _start ()
```

I now installed the missing debug symbols, and if the bug happens again, I will try to get a backtrace.

**Comment by RalfJung** [comment]
I think I managed to catch `globalkeys` when it was frozen:

```
#0  0x00007fbb105c9e2d in read () at ../sysdeps/unix/syscall-template.S:84
#1  0x0000000000420e20 in read (__nbytes=1, __buf=0x7ffd8f61c227, __fd=14) at /usr/include/x86_64-linux-gnu/bits/unistd.h:44
#2  readAll (fd=14, data=data@entry=0x7ffd8f61c227, length=length@entry=1)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/daemon/pipe_utils.cpp:59
#3  0x000000000040815c in Core::remoteXUngrabKey (this=this@entry=0x7ffd8f61d0f0, X11shortcut=...)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/daemon/core.cpp:1885
#4  0x0000000000408a83 in Core::remoteXGrabKey (this=0x7ffd8f61d0f0, X11shortcut=...)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/daemon/core.cpp:1826
#5  0x000000000040c51b in Core::grabOrReuseKey (this=this@entry=0x7ffd8f61d0f0, X11shortcut=..., shortcut=...)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/daemon/core.cpp:1907
#6  0x000000000040e9cd in Core::addOrRegisterClientAction (this=this@entry=0x7ffd8f61d0f0, shortcut=..., path=..., description=..., sender=...)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/daemon/core.cpp:2081
#7  0x00000000004126ef in Core::addClientAction (this=0x7ffd8f61d0f0, result=..., shortcut=..., path=..., description=..., sender=...)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/daemon/core.cpp:2135
#8  0x00000000004277ec in Core::qt_static_metacall (_o=_o@entry=0x7ffd8f61d0f0, _c=_c@entry=QMetaObject::InvokeMetaMethod, _id=_id@entry=3, _a=_a@entry=0x7ffd8f61c610)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/obj-x86_64-linux-gnu/daemon/moc_core.cpp:243
#9  0x00007fbb110d7f8a in QMetaObject::activate (sender=sender@entry=0x14b0ae0, signalOffset=<optimized out>, local_signal_index=local_signal_index@entry=0, 
    argv=argv@entry=0x7ffd8f61c610) at kernel/qobject.cpp:3713
#10 0x00007fbb110d8797 in QMetaObject::activate (sender=sender@entry=0x14b0ae0, m=m@entry=0x63bc60 <NativeAdaptor::staticMetaObject>, 
    local_signal_index=local_signal_index@entry=0, argv=argv@entry=0x7ffd8f61c610) at kernel/qobject.cpp:3578
#11 0x0000000000426a52 in NativeAdaptor::onAddClientAction (this=this@entry=0x14b0ae0, _t1=..., _t2=..., _t3=..., _t4=..., _t5=...)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/obj-x86_64-linux-gnu/daemon/moc_native_adaptor.cpp:272
#12 0x0000000000420405 in NativeAdaptor::addClientAction (this=0x14b0ae0, shortcut=..., path=..., description=..., id=@0x14b2400: 0)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/daemon/native_adaptor.cpp:43
#13 0x0000000000425a2b in OrgLxqtGlobalActionNativeAdaptor::addClientAction (id=<optimized out>, description=..., path=..., shortcut=..., this=0x14b6e00)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/obj-x86_64-linux-gnu/daemon/org.lxqt.global_key_shortcuts.native.cpp:39
#14 OrgLxqtGlobalActionNativeAdaptor::qt_static_metacall (_o=_o@entry=0x14b6e00, _id=_id@entry=0, _a=_a@entry=0x7ffd8f61c850, _c=QMetaObject::InvokeMetaMethod)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/obj-x86_64-linux-gnu/daemon/org.lxqt.global_key_shortcuts.native.moc:152
#15 0x0000000000425ca3 in OrgLxqtGlobalActionNativeAdaptor::qt_static_metacall (_a=0x7ffd8f61c850, _id=0, _c=QMetaObject::InvokeMetaMethod, _o=0x14b6e00)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/obj-x86_64-linux-gnu/daemon/org.lxqt.global_key_shortcuts.native.moc:199
#16 OrgLxqtGlobalActionNativeAdaptor::qt_metacall (this=0x14b6e00, _c=QMetaObject::InvokeMetaMethod, _id=0, _a=0x7ffd8f61c850)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/obj-x86_64-linux-gnu/daemon/org.lxqt.global_key_shortcuts.native.moc:200
#17 0x00007fbb12099913 in QDBusConnectionPrivate::deliverCall (this=this@entry=0x14a3650, object=object@entry=0x14b6e00, msg=..., metaTypes=..., slotIdx=5)
    at qdbusintegrator.cpp:978
#18 0x00007fbb1209d7c0 in QDBusConnectionPrivate::activateCall (this=this@entry=0x14a3650, object=0x14b6e00, flags=flags@entry=273, msg=...) at qdbusintegrator.cpp:892
#19 0x00007fbb1209e073 in QDBusConnectionPrivate::activateObject (this=0x14a3650, node=..., msg=..., pathStartPos=<optimized out>) at qdbusintegrator.cpp:1462
#20 0x00007fbb1209f26e in QDBusActivateObjectEvent::placeMetaCall (this=0x14c4e10) at qdbusintegrator.cpp:1580
#21 0x00007fbb110d8e61 in QObject::event (this=0x14b0ae0, e=<optimized out>) at kernel/qobject.cpp:1239
#22 0x00007fbb1145405c in QApplicationPrivate::notify_helper (this=this@entry=0x143f160, receiver=receiver@entry=0x14b0ae0, e=e@entry=0x14c4e10)
    at kernel/qapplication.cpp:3716
#23 0x00007fbb11459516 in QApplication::notify (this=0x7ffd8f61d0e0, receiver=0x14b0ae0, e=0x14c4e10) at kernel/qapplication.cpp:3499
#24 0x00007fbb110a95ab in QCoreApplication::notifyInternal (this=0x7ffd8f61d0e0, receiver=0x14b0ae0, event=event@entry=0x14c4e10) at kernel/qcoreapplication.cpp:965
#25 0x00007fbb110ab9a6 in QCoreApplication::sendEvent (event=0x14c4e10, receiver=<optimized out>)
    at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:224
#26 QCoreApplicationPrivate::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0, data=0x143f2d0) at kernel/qcoreapplication.cpp:1593
#27 0x00007fbb110abe88 in QCoreApplication::sendPostedEvents (receiver=receiver@entry=0x0, event_type=event_type@entry=0) at kernel/qcoreapplication.cpp:1451
#28 0x00007fbb110ff633 in postEventSourceDispatch (s=0x146e490) at kernel/qeventdispatcher_glib.cpp:271
#29 0x00007fbb0ecb9fd7 in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#30 0x00007fbb0ecba230 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#31 0x00007fbb0ecba2dc in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#32 0x00007fbb110ffa3f in QEventDispatcherGlib::processEvents (this=0x1480cd0, flags=...) at kernel/qeventdispatcher_glib.cpp:418
#33 0x00007fbb110a6d6a in QEventLoop::exec (this=this@entry=0x7ffd8f61cff0, flags=..., flags@entry=...) at kernel/qeventloop.cpp:204
#34 0x00007fbb110aee0c in QCoreApplication::exec () at kernel/qcoreapplication.cpp:1229
#35 0x00007fbb0f7d6c1c in QGuiApplication::exec () at kernel/qguiapplication.cpp:1539
#36 0x00007fbb11450495 in QApplication::exec () at kernel/qapplication.cpp:2976
#37 0x0000000000407139 in main (argc=1, argv=0x7ffd8f61d328) at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/daemon/main.cpp:251
```

**Comment by paulolieuthier** [comment]
Great. I'll take a look at that next week. Thanks!

**Comment by selairi** [comment]
lxqt-globalkeyd freezes when I start LXQt session and press any key shortcut. If I wait some minutes after session start, it works.

**Comment by selairi** [comment]
lxqt-globalkeysd only freezes at session start. I killed lxqt-globalkeysd and I restart it and it works OK. I can not freeze it again.

I have also tried to launch a plain openbox session. Then I run dbus-lauch and lxqt-globalkeysd, but everything it is OK.

**Comment by selairi** [comment]
I think that I found the bug:

lxqt-globalkeysd use a thread to talk with X11, I am going to call this thread "X11-lxqt-thread". Methods can talk with "X11-lxqt-thread" using pipes. Methods can register or unregister keyboard shortcuts using those pipes. In order to register a keyboard shortcut methods write orders in mX11RequestPipe pipe and wait the response from mX11ResponsePipe pipe.

But what happens if you try to run a key shortcut while and register a new shortcut, lxqt-globalkeys will stop, a deadlock.

You can see easily adding this code:

In daemon/core.cpp replace `Core::remoteXGrabKey(const X11Shortcut &X11shortcut)` by the next code:
```

bool Core::remoteXGrabKey(const X11Shortcut &X11shortcut)
{
    size_t X11Operation = X11_OP_XGrabKey;
    if (error_t error = writeAll(mX11RequestPipe[STDOUT_FILENO], &X11Operation, sizeof(X11Operation)))
    {
        log(LOG_CRIT, "Cannot write to X11 request pipe: %s", strerror(error));
        qApp->quit();
        return false;
    }
    if (error_t error = writeAll(mX11RequestPipe[STDOUT_FILENO], &X11shortcut.first, sizeof(X11shortcut.first)))
    {
        log(LOG_CRIT, "Cannot write to X11 request pipe: %s", strerror(error));
        qApp->quit();
        return false;
    }
    if (error_t error = writeAll(mX11RequestPipe[STDOUT_FILENO], &X11shortcut.second, sizeof(X11shortcut.second)))
    {
        log(LOG_CRIT, "Cannot write to X11 request pipe: %s", strerror(error));
        qApp->quit();
        return false;
    }
    sleep(1);
    printf("Grabkey...\n");
    wakeX11Thread();

    char signal;
    if (error_t error = readAll(mX11ResponsePipe[STDIN_FILENO], &signal, sizeof(signal)))
    {
        log(LOG_CRIT, "Cannot read from X11 response pipe: %s", strerror(error));
        qApp->quit();
        return false;
    }
    if (signal)
    {
        return false;
    }

    return true;
}

```

This code only sleep 1 second while lxqt-globalkeysd is adding a new shortcut. "Grabkey..." message will be shown when a new shortcut is adding.

Then kill lxqt-globalkeysd and launch this modified version from terminal. When "Grabkey..." message is shown start to press any keyboard shortcut. When that keyboard shortcut will be register lxqt-globalkeysd will stop forever, deadlock.

@RalfJung's debug shows:

```
#3  0x000000000040815c in Core::remoteXUngrabKey (this=this@entry=0x7ffd8f61d0f0, X11shortcut=...)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/daemon/core.cpp:1885
#4  0x0000000000408a83 in Core::remoteXGrabKey (this=0x7ffd8f61d0f0, X11shortcut=...)
    at /home/r/src/debian/lxqt/lxqt-globalkeys/build-area/lxqt-globalkeys-0.10.0/daemon/core.cpp:1826
``` 
It is trying to register and unregister a keyboard shortcut at the same time. "X11-lxqt-thread" could mixed both calls.

The solution: 

I don't know. Maybe we can learn how lxqt-globalkeysd use mutex and forgive:
- run keyboard shortcuts 
- add new shortcut
- remove shortcut.
can be run at the same time.

**Comment by selairi** [comment]
Hello,

I have just opened a new branch called ` test-mutex `. In this branch X11 key events are disabled until all keyboard shortcuts are registered (session start) or when a shortcut is being registered or deleted (session already started). It works OK for me.

Could @RalfJung be so kind as to test it?

**Comment by RalfJung** [comment]
The bug is not reproducible, it only happens roughly one in ten times. Testing does not help much. I will investigate the code to try and understand the issue, and the fix -- but I'm not sure when I will have the time.

**Comment by selairi** [comment]
I would like to merge `test-mutex` branch in `master` branch. It works OK for me.

If everybody is agreed with this, I will do it in a few days.

I would like to point out that lxqt-globalkeysd needs a revision. It uses threads, pipes and semaphores. Maybe [KGlobalAccel](http://api.kde.org/frameworks-api/frameworks5-apidocs/kglobalaccel/html/index.html) could help us.

**Comment by palinek** [comment]
> I would like to merge test-mutex branch in master branch. It works OK for me

Why didn't you make a PR?

**Comment by selairi** [comment]
I have found another bug. 

I remaped the keyboard of my laptop. The keyboard of my laptop is [like this one.](https://cdn6.bigcommerce.com/s-blhknq8/product_images/uploaded_images/acer-chromebook-11-keyboard-key-replacement.png?t=1425966466)

I use xmodmap in order to add a modifier key:

`xmodmap -e "keycode 133 = Mode_switch NoSymbol Super_L"`

Then I use a modifier key in order to add delete, home or print key. For example, the home key is added with:

`xmodmap -e "keycode 113 = Left NoSymbol Home NoSymbol Left Left"`

All keys work OK.

When I try to register print key (modifier key + F4 = Print key), this key is not recognised by lxqt-globalkeysd. Even so, xev shows me that Print key works OK.

But shift + Print (modifier key + F4 + Shift) is recognized by lxqt-globalkeysd.
